### PR TITLE
[Helper] Make TagFactory thread safe

### PR DESCRIPTION
**Pull request by alxbilger**
A mutex is added when trying to access a tag ID.

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by damienmarchal** [comment]
This class is using a common "ugly" design pattern often found in Sofa which mix 1) the actual object implementation, 2) is singleton exposure, and 3) a static API. 

My suggestion when we see this kind of pattern is to refactor it to separate the different concerns so that the thread safetyness is not mixed with implementation. 

Here is a suggested design that does not break the API (to much):
In the .h
```
/// Tag factory is only the public API forwarding to a totally private implementation. 
class TagFactory
{
public:    
    static unsigned int getID(std::string name);
   static std::string getName(unsigned int id);
};

In the .cpp
```cpp
/// Non thread safe class (and there is no reason why it should) because only the singletong should be protected.
class TagFactoryImpl  
{
    public:
         std::string getId(int id);
         void getName(); 
};

TagFactoryImpl::TagFactoryImpl::getInstance()
{
    /// ...
}

std::string TagFactory::getName()
{
             lock_mutex();
             getSingleton()->getName()
}

std::string Tag::Factory::getId()
{
             lock_mutex();
             getSingleton()->getId()    
}
```

File: Sofa/framework/Helper/src/sofa/helper/TagFactory.h [hunk 1]
```diff
@@ -19,57 +19,43 @@
...
 {
-protected:
+private:
 
-    /// the list of the tag names. the Ids are the indices in the vector
-    std::vector<std::string> tagsList;
+    static TagList* getTagList();
 
-    TagFactory();
+    static std::mutex s_mutex;
```
Comment (damienmarchal) [inline Sofa/framework/Helper/src/sofa/helper/TagFactory.h]: Why not having the mutex in the .cpp so it is totally hidden ?

Comment (alxbilger) [reply Sofa/framework/Helper/src/sofa/helper/TagFactory.h]: Because I did not think about it ;)

**Comment by damienmarchal** [comment]
Seems all right to me. 
I would votre for set to Ready

**Comment by epernod** [comment]
[ci-build][with-all-tests]

**Comment by damienmarchal** [comment]
This one is waiting for second review :)

**Review by damienmarchal — APPROVED** [review]
The refactoring of the class makes it more "private". 
And the use of a mutex is needed for thread safetiness. 

So all good to me.

File: Sofa/framework/Helper/test/TagFactory_test.cpp [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+TEST(TagFactory_test, addTag)
+{
+    EXPECT_EQ(TagFactory::getID("1"), 2);
+    EXPECT_EQ(TagFactory::getID("2"), 3);
+    EXPECT_EQ(TagFactory::getID("4"), 4);
+    EXPECT_EQ(TagFactory::getID("2"), 3);
+
+    EXPECT_EQ(TagFactory::getID("foo"), 5);
+    EXPECT_EQ(TagFactory::getID("bar"), 6);
+    EXPECT_EQ(TagFactory::getID("foo"), 5);
```
Comment (epernod) [inline Sofa/framework/Helper/test/TagFactory_test.cpp]: sorry I don't understand the test results.
Where is set the default list of Tag and what are the values?
In the first test I understand that the 14 first ids are pre-defined. But in the second test I understand that you are adding tags from scratch using getID

Comment (alxbilger) [reply Sofa/framework/Helper/test/TagFactory_test.cpp]: I think you would better understand if you read the code of the TagFactory.

- `getName`: gives the name corresponding to an ID. If the id is not yet in the factory, it returns an empty string.
- `getID`: gives the id corresponding to a name. If the name is not yet in the factory, an entry is created and the new id is returned.

The tag factory is initialized with 2 tags (not 14). It is initialized in this line: `TagList kTagList { {"0", "Visual"} };`

Comment (epernod) [reply Sofa/framework/Helper/test/TagFactory_test.cpp]: Ok this is the comment that confused me: `// id 14 is not yet in the tag list: an empty string is returned`
I thought it was the first empty entry.

File: Sofa/framework/Helper/test/TagFactory_test.cpp [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    EXPECT_EQ(TagFactory::getName(0), "0");
+    EXPECT_EQ(TagFactory::getName(1), "Visual");
+
+    // id 14 is not yet in the tag list: an empty string is returned
+    EXPECT_EQ(TagFactory::getName(14), "");
+}
+
+TEST(TagFactory_test, addTag)
+{
+    EXPECT_EQ(TagFactory::getID("1"), 2);
```
Comment (epernod) [inline Sofa/framework/Helper/test/TagFactory_test.cpp:39]:
```suggestion
    // Get id of existing Tags
    EXPECT_EQ(TagFactory::getID("0"), 0);
    EXPECT_EQ(TagFactory::getID("Visual"), 1);
    
    // Get id of new Tags. Will be added to the list
    EXPECT_EQ(TagFactory::getID("1"), 2);
```

File: Sofa/framework/Helper/test/TagFactory_test.cpp (hunk 2 above)
Comment (epernod) [reply Sofa/framework/Helper/test/TagFactory_test.cpp:45]: So maybe just a suggestion to test the getId on "0" and "Visual"

File: Sofa/framework/Helper/test/TagFactory_test.cpp (hunk 3 above)
Comment (epernod) [reply Sofa/framework/Helper/test/TagFactory_test.cpp:39]: So maybe just a suggestion to test getId on "0" and "Visual"

Comment (alxbilger) [reply Sofa/framework/Helper/test/TagFactory_test.cpp:39]: it's a relevant suggestion. Thanks

**Pull request merged**